### PR TITLE
Apple II a2bus Orange Micro Grappler Plus driver

### PR DESCRIPTION
**Pull request by goldnchild**
A driver for the Orange Micro Grappler Plus.

./mame64 apple2e -sl1 grapplerplus -sl1:grapplerplus:centronics ap2000

A simple test is to boot the apple 2, reset, then do PR#1   then type CTRL+I G enter.  This will do a graphic dump.

CTRL+I DG enter  is double size.
CTRL+I DERG enter is double size emphasized rotate graphic dump.
CTRL+I S will do a screen dump.

### CODE REVIEW

**Comment by rb6502** [comment]
This is super clean and I'm generally impressed, but two minor things:
1) Please put the DIP switches in numerical order for better readability
2) Use the m_rom lookup method from the Sider card (sider.cpp); it's a more modern way to do things.  I need to convert all the existing cards to do it so there's a better example for people.

Fix those and it's ready to go!

File: src/devices/bus/a2bus/grapplerplus.cpp [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+#include "grapplerplus.h"
+
+//**************************************************************************
+//  GLOBAL VARIABLES
+//**************************************************************************
+
+DEFINE_DEVICE_TYPE(A2BUS_GRAPPLERPLUS, a2bus_grapplerplus_device, "grapplerplus", "Grappler Interface Card")
+
+#define GRAPPLERPLUS_ROM_REGION  "grapplerplus_rom"
+#define GRAPPLERPLUS_CENTRONICS_TAG "centronics"
```
Comment (cuavas) [inline src/devices/bus/a2bus/grapplerplus.cpp:21]: You shouldn’t need a macro for the device tag – you can pass device finders in place of tags in machine configuration and address maps.

File: src/devices/bus/a2bus/grapplerplus.cpp [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+}
+
+void a2bus_grapplerplus_device::device_reset()
+{
+	m_started = true;
+	m_ack = 0;
+	m_irqenable = false;
+	m_irqbit = false;
+	m_autostrobe = false;
+	lower_slot_irq();
```
Comment (cuavas) [inline src/devices/bus/a2bus/grapplerplus.cpp:119]: Does the device actually have a latch that gets cleared when the slot reset line is asserted, or is this a workaround for something?

File: src/devices/bus/a2bus/grapplerplus.cpp [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+
+	// set initial state of the strobe line depending on the dipswitch
+	clear_strobe();
+}
+
+void a2bus_grapplerplus_device::device_timer(emu_timer &timer, device_timer_id tid, int param, void *ptr)
+{
+	clear_strobe();
+
+	m_timer->adjust(attotime::never);
```
Comment (cuavas) [inline src/devices/bus/a2bus/grapplerplus.cpp:130]: If you set the timer as a 1-shot, it won’t fire again – you shouldn’t need this.

File: src/devices/bus/a2bus/grapplerplus.cpp [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+			m_irqbit = false;
+			break;
+	}
+}
+
+WRITE_LINE_MEMBER( a2bus_grapplerplus_device::ack_w )
+{
+	if (m_started)
+	{
+		m_ack = state;
```
Comment (cuavas) [inline src/devices/bus/a2bus/grapplerplus.cpp:201]: You’re potentially losing the input state set by a connected device at start time by putting this inside the `if (m_started)`.

File: src/devices/bus/a2bus/grapplerplus.cpp [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+		{
+			raise_slot_irq();
+			m_irqbit = true;
+		}
+	}
+}
+
+WRITE_LINE_MEMBER( a2bus_grapplerplus_device::perror_w )
+{
+	m_perror = (state == ASSERT_LINE) ? true : false;
```
Comment (cuavas) [inline src/devices/bus/a2bus/grapplerplus.cpp:212]: You don’t need to use a ternary conditional that evaluates to itself – `(state == ASSERT_LINE)` already evaluates to `true` or `false`.

File: src/mame/drivers/apple2gs.cpp [hunk 6]
```diff
@@ -4593,6 +4594,7 @@ static void apple2_cards(device_slot_interface &device)
 	device.option_add("uthernet", A2BUS_UTHERNET);  /* A2RetroSystems Uthernet card */
 	device.option_add("sider2", A2BUS_SIDER2); /* Advanced Tech Systems / First Class Peripherals Sider 2 SASI card */
 	device.option_add("sider1", A2BUS_SIDER1); /* Advanced Tech Systems / First Class Peripherals Sider 1 SASI card */
+	device.option_add("grapplerplus", A2BUS_GRAPPLERPLUS); /* Orange Micro Grappler Plus Parallel Interface card */
```
Comment (cuavas) [inline src/mame/drivers/apple2gs.cpp:4597]: Should this also be added for the IIgs and possibly other Apple 2 families?  There are lists of slot options hiding all over the place for Apple II cards.

File: src/devices/bus/a2bus/grapplerplus.cpp (hunk 2 above)
Comment (goldnchild) [reply src/devices/bus/a2bus/grapplerplus.cpp:122]: I really don't know the exact answer.  I'm just copying the a2pic device where it makes sure the irq is lowered in the reset.

src/devices/bus/a2bus/a2bus.h:	void lower_slot_irq() { m_a2bus->set_irq_line(CLEAR_LINE, m_slot); }

File: src/mame/drivers/apple2gs.cpp (hunk 6 above)
Comment (goldnchild) [reply src/mame/drivers/apple2gs.cpp:4597]: I think I added it to the apple2, 2e and 2gs drivers at the end of the option_add list.

File: src/devices/bus/a2bus/grapplerplus.cpp (hunk 5 above)
Comment (goldnchild) [reply src/devices/bus/a2bus/grapplerplus.cpp]: ok, changed.

File: src/devices/bus/a2bus/grapplerplus.cpp (hunk 4 above)
Comment (goldnchild) [reply src/devices/bus/a2bus/grapplerplus.cpp]: ok, put before the if

File: src/devices/bus/a2bus/grapplerplus.cpp (hunk 3 above)
Comment (goldnchild) [reply src/devices/bus/a2bus/grapplerplus.cpp]: ok, removed

File: src/devices/bus/a2bus/grapplerplus.cpp (hunk 1 above)
Comment (goldnchild) [reply src/devices/bus/a2bus/grapplerplus.cpp]: I removed the macros and put the strings in directly.  If that's not good, can you point to an example of a device finder.  (I still don't know all the ins and outs of devices).

File: src/devices/bus/a2bus/grapplerplus.cpp [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+			break;
+
+		case 1: // Set rom bank to bank 2
+			m_rombank=1;
+			break;
+
+		case 2: // Reset interrupt request and IRQ data bit
+			m_irqenable = false;
+			m_irqbit = false;
+			break;
```
Comment (cuavas) [inline src/devices/bus/a2bus/grapplerplus.cpp:184]: If this is supposed to acknowledge an interrupt request, you need to deassert it as well.

File: src/devices/bus/a2bus/grapplerplus.cpp [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+{
+	m_ack = state;
+	if (m_started)
+	{
+		if (state && m_irqenable)
+		{
+			raise_slot_irq();
+			m_irqbit = true;
+		}
+	}
```
Comment (cuavas) [inline src/devices/bus/a2bus/grapplerplus.cpp:205]: I just checked the Centronics port device, and it doesn’t detect changes in state for you.  You’ll need to detect edges yourself if the card does this.

**Review by cuavas — COMMENTED** [review]
There seem to be multiple versions of this card, known as “long” and “short” versions.
* https://i0.wp.com/www.applerescueofdenver.com/wp-content/uploads/grappler-plus.jpg
* https://i1.wp.com/www.applerescueofdenver.com/wp-content/uploads/grappler-plus-2.jpg

There also seems to have been a version with the Grappler+ and Bufferboard functionality integrated on a single card.

At a guess:
* U2 (74LS374) is the data output buffer
* U4 is obviously the ROM
* U8 (74LS74) is used to latch interrupt status, etc.
* U9 (74LS373) buffers the status inputs and DIP switches
* The other chips are probably glue logic and the ROM bank latch

Unfortunately there don’t seem to be photos that will allow tracing where all the signals go, and I can’t find a schematic.

**Comment by cuavas** [comment]
Of course, after saying that, I find the manual and schematics for the Grappler+ and Buffered Grappler+ (page 14 has the schematic): https://ia600704.us.archive.org/18/items/Grappler_Operators_Manual/Grappler_Operators_Manual.pdf

So some feedback to make the emulation more accurate:
* The DIP switch for selecting between 7-bit and 8-bit mode applies after the output buffer.  It forces D7 low whenever it’s set.  To emulate this properly you’ll need an input changed handler to detect changes and update D7 if it’s changed on-the-fly.
* The strobe duration is actually derived from the phase 1 clock input – see U5A/U5B.
* The Q output of U2C will be high whenever the /RST input (from the bus) or the /ACK input (from the printer) is low.  When neither of these inputs are low, it can be reset by writing to the data latch.
* Enable/disable interrupt doesn’t affect the latched acknowledge status, it just gates the interrupt.  If it’s re-enabled while still latched, the interrupt will be asserted again immediately.
* The latched acknowledge value somehow affects the A6 input to the ROM – see how it’s fed to U3B.

There’s a bit more work to get this all sorted out, I think.  The schematic is there at least, so the logic equations can be worked out.

**Comment by cuavas** [comment]
OK, let’s start looking at the address decoding for the c0nx range:

Writes to c0nx when A0 and A1 are both low:
* Latch data in output buffer
* Clear acknowledge latch
* After a delay of one phase 1 clock cycle, assert /STROBE for one phase 1 clock cycle

Writes to c0nx when A0 is high:
* Select ROM bank 1

Writes to c0nx when A1 is high:
* Disable interrupt (will not clear acknowledge latch)

Writes to x0nx when A2 is high:
* Enable interrupt (may cause interrupt to be asserted immediately)

Note that due to incomplete address decoding, several things can be triggered by a single write.  For example you can write data and enable interrupt in the same write by using offset 4, or you can enable interrupt and select ROM bank 1 using offset 5.  The state of the interrupt enable latch will be indeterminate if you use offset 6.

**Comment by cuavas** [comment]
As you’ve correctly determined, any read of the c0nx reads status.  The bits from zero to seven are:
* Acknowledge latch (note that this isn’t affected by whether interrupts are enabled, but it is latched, and reset sets the latch)
* Select
* Paper empty
* Busy
* DIP switch 4
* DIP switch 3
* DIP switch 2
* Interrupt state (1 if interrupt is asserted)

**Comment by cuavas** [comment]
The acknowledge latch is set either the reset input from the bus being asserted, or /ACK from the printer being asserted (low).  It’s a simple SR latch – there’s no edge detection.  If the printer keeps /ACK asserted while you write data, it won’t be cleared.  You’ll need to keep track of the /ACK input so you don’t clear it when it wouldn’t be cleared in reality.

**Comment by cuavas** [comment]
The ROM addressing is a bit weird, let’s look at each bit individually, from A0 to A11 (ACK is the acknowledge latch, other signals are from the bus):
* A0
* A1
* A2
* A3
* A4
* A5
* A6 && !(A7 && !/IOSEL && !ACK)
* A7
* A8 && /IOSEL
* A9 && /IOSEL
* A10 && /IOSEL
* bank latch

Accesses to cnxx reset the bank latch and can only access the low 256 bytes of the ROM.  However, there’s the weirdness with ROM A6 being gated using A7, the acknowledge latch and /IOSEL.  If A7 is high, and it’s a cnxx access and the ACK latch is clear, ROM A6 will be forced low.

**Comment by cuavas** [comment]
The reset input from the bus sets the acknowledge latch and disables the interrupt, so it will have the effect of deasserting the interrupt output.  It doesn’t affect the selected ROM bank or the strobe output.

**Comment by goldnchild** [comment]
The DIP switch for selecting between 7-bit and 8-bit mode applies after the output buffer. It forces D7 low whenever it’s set. To emulate this properly you’ll need an input changed handler to detect changes and update D7 if it’s changed on-the-fly.

Would it be considered wrong if I just did:

  m_ctx_data_out->write(data & (m_dsw1->read() & 0x01 ? 0xff : 0x7f));

**Comment by cuavas** [comment]
> Would it be considered wrong if I just did:
> 
> m_ctx_data_out->write(data & (m_dsw1->read() & 0x01 ? 0xff : 0x7f));

That's would simulate the DIP switch applying before the latch, but it applies after the latch. It actually latches all eight data bits. If someone flips the DIP switch while bit 7 is latched high, the D7 output will change immediately.

**Comment by rb6502** [comment]
@goldnchild You want PORT_CHANGED_MEMBER to get a callback when the dip switch is changed.  It's mostly a pointless distinction given that the window before the printer latches the data bits on the strobe assertion is measured in nanoseconds, but it's not too painful to do in this case and it's kind of bragging rights.  Look at AJR's byte8251.cpp Apple II card for an example.

On the $C0nX writes, replace the switch as follows:
if ((offset & 3) == 0) for the A0 and A1 both low case, if (offset & 1) for A0 high, if (offset & 2) for A1 high, if (offset & 4) for A2 high.  When you do it like this you'll automatically get the mirroring the schematics describe.

For the $CnXX A6 weirdness, if ((offset & 0x80) && (not /ACK)) then offset &= ~0x40;  This is a Woz trick from the original Apple II parallel card that I'm kind of surprised to find on a board with a much higher gate budget.

**Comment by cuavas** [comment]
I implemented it pretty much straight off the schematics, and dug up another ROM revision: 515e85ba80a320981f5f4f2546f73ac06cd6db34

It seemed to work in Epson mode, but I haven’t tested it extensively.

**Comment by goldnchild** [comment]
Thanks, Vas!   I've been trying to digest it all, but I must admit that I'm slow 8-)

**Comment by Tafoid** [comment]
Not sure what is causing it specifically and the machine is known not to work, but there was a display change I should note only in this one machine in the whole driver I tested.

apple2gsr3p
apple2gsr3p-diff

**Pull request closed without merging**